README.md
# koa-exception-handler

<a href="https://circleci.com/gh/shuidi-fed/koa-exception-handler/tree/master"><img src="https://img.shields.io/circleci/build/github/shuidi-fed/koa-exception-handler/master.svg" alt="build status"/></a>
[![](https://img.shields.io/npm/v/koa-exception-handler.svg)](https://www.npmjs.com/package/koa-exception-handler)
<a href="https://github.com/shuidi-fed/koa-exception-handler"><img src="https://img.shields.io/github/license/shuidi-fed/koa-exception-handler.svg" alt="License"/></a>
<a href="http://commitizen.github.io/cz-cli/"><img src="https://img.shields.io/badge/commitizen-friendly-brightgreen.svg" alt="Commitizen friendly"/></a>

> Customizable koa exception (error) handling middleware

[**中文文档**](./README.zh-CN.md)

## Installation

```sh
yarn add koa-exception-handler
```

## Usage

```js
const Koa = require('koa')
const koaException = require('koa-exception-handler')

const app = new Koa()

app.use(koaException(options))
```

## options

### options.errorLogger

Customize `errorLogger` function.

- Type: `function`
- Default: `console.error`

### options.production

Specifies whether it is a production environment that controls the behavior of the middleware when an error occurs on the server side.

- Type: `boolean`
- Default: `process.env.NODE_ENV !== 'development'`

**When `options.production` is `true`, if an error occurs on the server, the `options.fallback` function is executed without sending an error to the client. Of course, you can implement the behavior you expect in `options.fallback`.**

### optioins.clientErrorHandler

The handler that is executed when an error occurs on the client.

- Type: `function`
- Default: `undefined`
- arguments:
  - ctx `Koa context`
  - err `error` - Instance of `Error`

- example:

```js
app.use(koaException({
  clientErrorHandler: (ctx, err) => {
    ctx.status = err.code
    ctx.type = 'text/plain; charset=utf-8'
    ctx.body = 'Client error: ' + err.toString()
  }
}))
```

### optioins.serverErrorHandler

In a **non-production** environment, the handler executed when an error occurs on the server side.

- Type: `function`
- Default: `undefined`
- arguments:
  - ctx `Koa context`
  - err `error` - Instance of `Error`

Usage reference `optioins.clientErrorHandler`.

### optioins.fallback

Similar to `optioins.serverErrorHandler`, except that `optioins.fallback` is a handler that is executed in the production environment when an error occurs on the server side.

- Type: `function`
- Default: `undefined`
- arguments:
  - ctx `Koa context`
  - err `error` - Instance of `Error`

CHANGELOG.md
## <small>0.1.1 (2019-07-18)</small>

* :beetle: Modify error message ([173fd91](https://github.com/shuidi-fed/koa-exception/commit/173fd91))
* :books: errorLogger ([3960bee](https://github.com/shuidi-fed/koa-exception/commit/3960bee))
* :books: installation ([e188e57](https://github.com/shuidi-fed/koa-exception/commit/e188e57))
* :books: Translate to zh-CN ([8ec8675](https://github.com/shuidi-fed/koa-exception/commit/8ec8675))
* :nut_and_bolt: Add license ([0291f40](https://github.com/shuidi-fed/koa-exception/commit/0291f40))
* :nut_and_bolt: tweak ([4a6cd32](https://github.com/shuidi-fed/koa-exception/commit/4a6cd32))
* :police_car: passWithNoTests ([0689cee](https://github.com/shuidi-fed/koa-exception/commit/0689cee))
* :tada: init ([6fee7b2](https://github.com/shuidi-fed/koa-exception/commit/6fee7b2))
* :traffic_light: Add a build status badge ([7801b0c](https://github.com/shuidi-fed/koa-exception/commit/7801b0c))
* :traffic_light: config file ([afb11b7](https://github.com/shuidi-fed/koa-exception/commit/afb11b7))





lib/__test__/index.spec.js
const KoaException = require('../index')

test('Handling client errors', async () => {
  let err = null
  const next = async () => {
    err = new Error()
    err.code = 400
    throw err
  }

  const options = {
    errorLogger: jest.fn(),
    clientErrorHandler: jest.fn(ctx => {
      ctx.status = 400
      ctx.type = 'text/plain; charset=utf-8'
      ctx.body = 'Client error'
    })
  }

  const mw = KoaException(options)

  const ctx = {}

  await mw(ctx, next)

  const loggerCalls = options.errorLogger.mock.calls
  const loggerArgs = loggerCalls[0]
  const clientErrorHandlerCalls = options.clientErrorHandler.mock.calls
  const clientErrorHandlerArgs = options.clientErrorHandler.mock.calls[0]

  expect(loggerCalls.length).toBe(1)
  expect(loggerArgs.length).toBe(3)
  expect(clientErrorHandlerCalls.length).toBe(1)
  expect(clientErrorHandlerArgs.length).toBe(2)
  expect(clientErrorHandlerArgs[1]).toBe(err)
  expect(loggerArgs[0]).toBe('[CLIENT ERROR]')
  expect(loggerArgs[1]).toBe(err)
  expect(loggerArgs[2]).toEqual(clientErrorHandlerArgs[0])
})

test('Handling server errors - development', async () => {
  let err = null
  const next = async () => {
    err = new Error()
    err.code = 500
    throw err
  }

  const options = {
    errorLogger: jest.fn(),
    production: false,
    serverErrorHandler: jest.fn(ctx => {
      ctx.status = 500
      ctx.type = 'text/plain; charset=utf-8'
      ctx.body = 'Server error'
    })
  }

  const mw = KoaException(options)

  const ctx = {}

  await mw(ctx, next)

  const loggerCalls = options.errorLogger.mock.calls
  const loggerArgs = loggerCalls[0]
  const serverErrorHandlerCalls = options.serverErrorHandler.mock.calls
  const serverErrorHandlerArgs = options.serverErrorHandler.mock.calls[0]

  expect(loggerCalls.length).toBe(1)
  expect(loggerArgs.length).toBe(3)
  expect(serverErrorHandlerCalls.length).toBe(1)
  expect(serverErrorHandlerArgs.length).toBe(2)
  expect(serverErrorHandlerArgs[1]).toBe(err)
  expect(loggerArgs[0]).toBe('[SERVER ERROR]')
  expect(loggerArgs[1]).toBe(err)
  expect(loggerArgs[2]).toEqual(serverErrorHandlerArgs[0])
})

test('Handling server errors - production', async () => {
  let err = null
  const next = async () => {
    err = new Error()
    err.code = 500
    throw err
  }

  const options = {
    errorLogger: jest.fn(),
    serverErrorHandler: jest.fn(ctx => {
      ctx.status = 500
      ctx.type = 'text/plain; charset=utf-8'
      ctx.body = 'Server error'
    }),
    fallback: jest.fn(ctx => {
      ctx.fallback = true
    })
  }

  const mw = KoaException(options)

  const ctx = {}

  await mw(ctx, next)

  const loggerCalls = options.errorLogger.mock.calls
  const loggerArgs1 = loggerCalls[0]
  const loggerArgs2 = loggerCalls[1]
  const serverErrorHandlerCalls = options.serverErrorHandler.mock.calls
  const fallbackCalls = options.fallback.mock.calls
  const fallbackArgs = fallbackCalls[0]

  expect(loggerCalls.length).toBe(2)
  expect(loggerArgs1.length).toBe(3)
  expect(loggerArgs1[0]).toBe('[SERVER ERROR]')
  expect(loggerArgs1[1]).toBe(err)
  expect(loggerArgs1[2]).toBe(ctx)
  expect(loggerArgs2.length).toBe(1)
  expect(loggerArgs2[0]).toBe('FALLBACK')
  expect(serverErrorHandlerCalls.length).toBe(0)

  expect(fallbackCalls.length).toBe(1)
  expect(fallbackArgs.length).toBe(2)
  expect(fallbackArgs[0].fallback).toBe(true)
  expect(fallbackArgs[1]).toBe(err)
})


lib/index.js

/**
 * According to the http status code to determine whether it is a client request error
 *
 * @param {number} code
 * @return {boolean}
 */
const isClientError = exports.isClientError = code => code >= 400 && code < 500

/**
 * @param {object} [options]
 * @param {functoin} [options.logger] Custom logger function
 * @param {boolean} [options.production] Specify whether it is a production environment
 * @param {function} [options.clientErrorHandler] Execute when a client request error occurs
 * @param {function} [options.serverErrorHandler] Execute when a server error occurs
 * @param {function} [options.fallback] Executed when an error occurs in the server of the production environment
 */
module.exports = options => {
  const {
    errorLogger = console.error,
    clientErrorHandler,
    serverErrorHandler,
    fallback
  } = options
  let { production } = options

  production = (production != undefined && production != null) // eslint-disable-line
    ? production
    : process.env.NODE_ENV !== 'development'

  return async (ctx, next) => {
    try {
      await next()
    } catch (err) {
      // Handling client errors
      if (isClientError(err.code)) {
        errorLogger('[CLIENT ERROR]', err, ctx)

        if (clientErrorHandler) {
          clientErrorHandler(ctx, err)
        } else {
          responseError(ctx, err.code, err.toString())
        }
        return
      }

      errorLogger('[SERVER ERROR]', err, ctx)

      // Handling server errors
      if (!production) {
        if (serverErrorHandler) {
          serverErrorHandler(ctx, err)
        } else {
          responseError(ctx, err.code || 500, err.toString())
        }
        return
      }

      // Production environment fallback
      if (fallback) {
        errorLogger('FALLBACK')
        fallback(ctx, err)
      }
    }
  }
}

function responseError (ctx, status, errorString) {
  ctx.status = status
  ctx.type = 'text/plain; charset=utf-8'
  ctx.body = `${isClientError(status) ? 'Client' : 'Server'} error: ` + errorString
}
